File: sdk/python/neura/state.py
"""State operations — persistent key-value storage."""

from typing import Any, Optional
from .client import HttpClient


class StateAPI:
    """High-level state operations."""

    def __init__(self, client: HttpClient):
        self._client = client

    def set(self, key: str, value: Any, idempotency_key: Optional[str] = None) -> dict:
        """Set a key-value state entry."""
        result = self._client.request(
            "POST", "/api/state", {"key": key, "value": value}, idempotency_key
        )
        return result["data"]

    def get(self, key: str) -> dict:
        """Get a state entry by key."""
        result = self._client.request("GET", f"/api/state/{__import__('urllib.parse').parse.quote(key)}")
        return result["data"]

    def list(self) -> list[dict]:
        """List all state entries."""
        result = self._client.request("GET", "/api/state")
        return result["data"]

    def delete(self, key: str) -> None:
        """Delete a state entry."""
        self._client.request("DELETE", f"/api/state/{__import__('urllib.parse').parse.quote(key)}")

File: sdk/typescript/src/state.ts
// ============================================================
// Neura SDK — State Operations
// ============================================================

import { HttpClient } from './client'
import type { StateEntry } from './types'

export class StateAPI {
  constructor(private client: HttpClient) {}

  /**
   * Set a key-value state entry.
   * Overwrites any existing value for the same key.
   * 
   * @example
   * await neura.state.set('current_goal', { task: 'Build API', priority: 'high' })
   */
  async set(key: string, value: unknown, idempotencyKey?: string): Promise<StateEntry> {
    const { data } = await this.client.request<StateEntry>('POST', '/api/state', { key, value }, {
      idempotencyKey,
    })
    return data
  }

  /**
   * Get a specific state value by key.
   * Throws NeuraHttpError with code 'not_found' if key doesn't exist.
   */
  async get(key: string): Promise<StateEntry> {
    const { data } = await this.client.request<StateEntry>('GET', `/api/state/${encodeURIComponent(key)}`)
    return data
  }

  /**
   * List all state entries for this agent.
   */
  async list(): Promise<StateEntry[]> {
    const { data } = await this.client.request<StateEntry[]>('GET', '/api/state')
    return data
  }

  /**
   * Delete a state entry by key.
   */
  async delete(key: string): Promise<void> {
    await this.client.request('DELETE', `/api/state/${encodeURIComponent(key)}`)
  }
}
